src/imperative/imperative.cc

```diff
@@ -316,181 +312,223 @@ std::vector<NDArray*> Imperative::Backward(
       info.outputs.back() = static_cast<real_t>(1.0);
     }
   }
+  return ograd_entries;
+}
 
-  // Get gradient graph
-  Symbol sym;
-  sym.outputs = graph.outputs;
-  std::vector<NodeEntry> xs;
-  std::vector<NDArray*> x_grads;
-  std::vector<OpReqType> x_reqs;
-  if (variables.size()) {
-    xs.reserve(variables.size());
-    x_grads.reserve(variables.size());
-    x_reqs.reserve(variables.size());
+struct Imperative::GradientVariableNodes {
```

Comment:
Since these vector will always have the same size, would it make sense to create a struct like the following:
```suggestion
struct Imperative::GradientVariableNode {
     nnvm::NodeEntry variable_node;
     NDArray *gradient
     OpReqType op_req_type 
```
and return a vector of this structure from `CreateGradientVariableNodes` function?

Reply:
This is the old SOA vs AOS dilema, in general is less efficient to do AOS.
https://www.youtube.com/watch?v=Dj3lkfIEUyA
This is well known in gaming circles, as you loose quite a bit of cache efficiency due to AoS.

I don't see a real benefit of having a single array here.

Reply:
The benefit of keeping an array of struct is to make sure they have the same length. In your design, there is no direct way to guarantee this invariance.

Reply:
The efficiency of AOS vs SOA does not apply here as you only perform adding element and no ALU related ops in the loop.

On the other hand, using a struct of vectors seems only move the original code to a struct, could we improve the readability during this refactoring?